org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/DefaultGradleBuild.java

```diff
@@ -84,13 +85,15 @@ public BuildConfiguration getBuildConfig() {
     @Override
     public BuildLauncher newBuildLauncher(RunConfiguration runConfiguration, TransientRequestAttributes transientAttributes) {
         // TODO (donat) once GradleWorkspaceManager#getGradleBuild(FixedRequestAttributes) is removed then we should only allow run config that contain the same build config
-        return ConnectionAwareLauncherProxy.newBuildLauncher(runConfiguration.toRequestAttributes(), transientAttributes);
+        BuildEnvironment buildEnvironment = this.modelProvider.reloadBuildEnvironment(transientAttributes);
```

Comment:
This should happen in the `ConnectionAwareLauncherProxy` instead. `GradleBuild` and `ModelProvider` shouldn't need to know about this.